lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,953 @@
+			'block_json' => array( '__experimentalTextDecoration' ),
+		),
+		'textTransform'            => array(
+			'theme_json' => array( 'typography', 'textTransform' ),
+			'block_json' => array( '__experimentalTextTransform' ),
+		),
+	);
+
+	/**
+	 * Constructor.
+	 *
+	 * @param array $contexts A structure that follows the theme.json schema.
+	 */
+	public function __construct( $contexts = array() ) {
+		$this->contexts = array();
+
+		if ( ! is_array( $contexts ) ) {
+			return;
+		}
+
+		$metadata = $this->get_blocks_metadata();
+		foreach ( $contexts as $key => $context ) {
+			if ( ! array_key_exists( $key, $metadata ) ) {
+				// Skip incoming contexts that can't be found
+				// within the contexts registered.
+				continue;
+			}
+
+			// Filter out top-level keys that aren't valid according to the schema.
+			$context = array_intersect_key( $context, self::SCHEMA );
+
+			// Selector & Supports are always taken from metadata.
+			$this->contexts[ $key ]['selector'] = $metadata[ $key ]['selector'];
+			$this->contexts[ $key ]['supports'] = $metadata[ $key ]['supports'];
+
+			// Process styles subtree.
+			$this->process_key( 'styles', $context, self::SCHEMA );
+			if ( array_key_exists( 'styles', $context ) ) {
+				$this->process_key( 'color', $context['styles'], self::SCHEMA['styles'] );
+				$this->process_key( 'typography', $context['styles'], self::SCHEMA['styles'] );
+
+				if ( 0 === count( $context['styles'] ) ) {
+					unset( $context['styles'] );
+				} else {
+					$this->contexts[ $key ]['styles'] = $context['styles'];
+				}
+			}
+
+			// Process settings subtree.
+			$this->process_key( 'settings', $context, self::SCHEMA );
+			if ( array_key_exists( 'settings', $context ) ) {
+				$this->process_key( 'color', $context['settings'], self::SCHEMA['settings'] );
+				$this->process_key( 'spacing', $context['settings'], self::SCHEMA['settings'] );
+				$this->process_key( 'typography', $context['settings'], self::SCHEMA['settings'] );
+
+				if ( 0 === count( $context['settings'] ) ) {
+					unset( $context['settings'] );
+				} else {
+					$this->contexts[ $key ]['settings'] = $context['settings'];
+				}
+			}
+		}
+	}
+
+	/**
+	 * Returns the metadata for each block.
+	 *
+	 * Example:
+	 *
+	 * {
+	 *   'global': {
+	 *     'selector': ':root'
+	 *     'supports': [ 'fontSize', 'backgroundColor' ],
+	 *     'blockName': 'global',
+	 *   },
+	 *   'core/heading/h1': {
+	 *     'selector': 'h1'
+	 *     'supports': [ 'fontSize', 'backgroundColor' ],
+	 *     'blockName': 'core/heading',
+	 *   }
+	 * }
+	 *
+	 * @return array Block metadata.
+	 */
+	public function get_blocks_metadata() {
+		if ( null !== self::$blocks_metadata ) {
+			return self::$blocks_metadata;
+		}
+
+		self::$blocks_metadata = array();
+
+		$registry = WP_Block_Type_Registry::get_instance();
+		$blocks   = array_merge(
+			$registry->get_all_registered(),
+			array(
+				'global' => new WP_Block_Type(
```

Comment:
Would it make sense to move the global support definitions to constant, instead of having them in the constructor?

Reply:
We can do so, yup https://github.com/WordPress/gutenberg/pull/26803/commits/d82db958e03204a7c381d0b3e1dbea2e086575a1